### PR TITLE
fix: Emailverification send email functions with multitenancy

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Emailverification send email functions with multitenancy

## Related issues

-   https://github.com/supertokens/supertokens-node/pull/625

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [x] Make sure that `syncio` / `asyncio` functions are consistent.

### CODE REVIEW

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 1]
```diff
@@ -146,3 +157,76 @@ async def send_email(
     return await EmailVerificationRecipe.get_instance().email_delivery.ingredient_interface_impl.send_email(
         input_, tenant_id or DEFAULT_TENANT_ID, user_context
     )
+
+
+async def create_email_verification_link(
+    user_id: str,
+    email: Optional[str],
+    tenant_id: Optional[str],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:165]: set it to None by default

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 2]
```diff
@@ -146,3 +157,76 @@ async def send_email(
...
+            recipe_instance.get_recipe_id(),
+            tenant_id or DEFAULT_TENANT_ID,
+        )
+    )
+
+
+async def send_email_verification_email(
+    user_id: str,
+    email: Optional[str],
+    tenant_id: Optional[str],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:195]: Set it to None by default

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 3]
```diff
@@ -146,3 +157,76 @@ async def send_email(
...
+    ):
+        return SendEmailVerificationEmailAlreadyVerifiedError()
+
+    await send_email(
+        VerificationEmailTemplateVars(
+            user=VerificationEmailTemplateVarsUser(user_id, email),
+            email_verify_link=email_verification_link.link,
+            user_context={},
+        ),
+        tenant_id,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:228]: tenant_id should be part of template vars

File: supertokens_python/recipe/emailverification/syncio/__init__.py [hunk 4]
```diff
@@ -88,3 +88,29 @@ def send_email(
     from supertokens_python.recipe.emailverification.asyncio import send_email
 
     return sync(send_email(input_, tenant_id, user_context))
+
+
+def create_email_verification_link(
+    user_id: str,
+    email: Optional[str],
+    tenant_id: Optional[str],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/syncio/__init__.py:96]: set to none

File: supertokens_python/recipe/emailverification/syncio/__init__.py [hunk 5]
```diff
@@ -88,3 +88,29 @@ def send_email(
...
+        create_email_verification_link,
+    )
+
+    return sync(create_email_verification_link(user_id, email, tenant_id, user_context))
+
+
+def send_email_verification_email(
+    user_id: str,
+    email: Optional[str],
+    tenant_id: Optional[str],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/syncio/__init__.py:109]: set to none

File: supertokens_python/recipe/dashboard/interfaces.py [hunk 6]
```diff
@@ -251,6 +251,13 @@ def to_json(self) -> Dict[str, Any]:
         return {"status": self.status}
 
 
+class UserEmailVerifyTokenPostAPIEmailDoesntExistErrorResponse(APIResponse):
+    status: str = "EMAIL_ALREADY_VERIFIED_ERROR"
+
+    def to_json(self) -> Dict[str, Any]:
+        return {"status": self.status}
+
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/dashboard/interfaces.py:259]: this error is not needed

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 7]
```diff
@@ -30,15 +30,42 @@
...
+
+class CreateEmailVerificationLinkOkResult:
+    status = "OK"
+
+    def __init__(self, link: str):
+        self.link = link
+
+
+class SendEmailVerificationEmailDoesntExistError:
+    status = "EMAIL_DOES_NOT_EXIST_ERROR"
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:55]: this error is not needed

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 8]
```diff
@@ -146,3 +157,76 @@ async def send_email(
...
+
+
+async def send_email_verification_email(
+    user_id: str,
+    email: Optional[str],
+    tenant_id: Optional[str],
+    user_context: Dict[str, Any],
+) -> Union[
+    SendEmailVerificationEmailOkResult,
+    SendEmailVerificationEmailDoesntExistError,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:199]: this error is not needed

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 9]
```diff
@@ -146,3 +157,76 @@ async def send_email(
...
+    SendEmailVerificationEmailAlreadyVerifiedError,
+]:
+    if email is None:
+        recipe_instance = EmailVerificationRecipe.get_instance()
+
+        email_info = await recipe_instance.get_email_for_user_id(user_id, user_context)
+        if isinstance(email_info, GetEmailForUserIdOkResult):
+            email = email_info.email
+        elif isinstance(email_info, EmailDoesNotExistError):
+            return SendEmailVerificationEmailDoesntExistError()
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:209]:
```suggestion
            return SendEmailVerificationEmailAlreadyVerifiedError()
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/dashboard/interfaces.py (hunk 6 above)
Comment (KShivendu) [reply supertokens_python/recipe/dashboard/interfaces.py:259]: removed

File: supertokens_python/recipe/emailverification/asyncio/__init__.py (hunk 8 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/asyncio/__init__.py]: Removed

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**